R/render-cached-plot.R

```diff
@@ -0,0 +1,404 @@
+#'     there is potentially a large performance benefit if there are many users
+#'     of the application. If plots cannot be safely shared across users, this
+#'     should not be used.}
+#'  }
+#'
+#'   In some cases, you may want to manually specify the cache directory. This
+#'   can be useful if you want the cache to persist across multiple runs of an
+#'   application, or even across multiple R processes.
+#'
+#' \describe{
+#'   \item{3}{To have the cache persist across multiple runs of an R process,
+#'     use \code{scope=file.path(dirname(tempdir()), "plot1_cache")}.
+#'     This will create a subdirectory in your system temp directory named
+#'     \code{plot1_cache} (where \code{plot1_cache} is replaced with a unique
+#'     name of your choosing). When the R process exits, it will automatically
+#'     be removed.}
+#'   \item{4}{To have the cache persist even across multiple R processes, you
+#'     can set \code{cacheDir} to a location outside of the temp directory.
+#'     For example, it could be a subdirectory of the application, as in
+#'     \code{scope="plot1_cache"}}.
+#' }
+#'
+#'   Please note that if you specify a directory, that directory should only be
+#'   used to plot cache files. If it contains any other files or directories,
+#'   they could be removed when the cache is invalidated. Additionally, the
+#'   directory will not automatically be cleaned up or removed when the Shiny
+#'   application exits.
+#'
+#' @inheritParams renderPlot
+#' @param cacheKeyExpr An expression that returns a cache key. This key should
+#'   be a unique identifier for a plot: the assumption is that if the cache key
+#'   is the same, then the plot will be the same.
+#' @param cacheResetEventExpr An expression or block of code that accesses any
+#'   reactives whose invalidation should cause the cached plots to be cleared.
+#'   If \code{NULL} (the default) the cache will not get cleared.
+#' @param baseWidth A base value for the width of the cached plot.
+#' @param aspectRatioRate A multiplier for different possible aspect ratios. For
+#'   example, with a value of 1.2, the possible aspect ratios for plots will be
+#'   1:1, 1:1.2, 1:1.44, and so on, getting wider, as well as 1.2:1, 1.44:1, and
+#'   so on, getting taller.
+#' @param growthRate A multiplier for different cached image sizes. For example,
+#'   with a \code{width} of 400 and a \code{growthRate} of 1.25, there will be
+#'   possible cached images of widths 256, 320, 400, 500, 625, and so on, both
+#'   smaller and larger.
+#' @param res The resolution of the PNG, in pixels per inch.
+#' @param scope The scope of the cache. This can be \code{"app"} (the default),
+#'   \code{"session"}, or the path to a directory to store cached plots. See the
+#'   Cache Scoping section for more information.
+#'
+#' @export
+renderCachedPlot <- function(expr, cacheKeyExpr, cacheResetEventExpr = NULL,
+  baseWidth = 400, aspectRatioRate = 1.2, growthRate = 1.2, res = 72,
```

Comment:
It'd be cool to replace baseWidth/aspectRatioRate/growthRate arguments with a single argument that's a function that takes width/height and returns width/height. Then you can have multiple functions that we ship that represent different policies.